text/0083-rawkv-cross-cluster-replication.md

```diff
@@ -0,0 +1,183 @@
+
+In this proposal, we provide a TiKV sink with configurable concurrency, batch size, and backoff parameters, to be adapted to different network environments.
+
+## Detailed Design
+
+### 1. Utilize TiKV-BR & TiKV-CDC as Data Replication Component
+
+**TiKV-BR**, a fork of [BR](https://github.com/pingcap/tidb/tree/master/br), is the tool for backup and restoration of TiKV. We utilize **TIKV-BR** to accomplish data initialization of recovery cluster.
+
+**TiKV-CDC**, a fork of [TiCDC](https://github.com/pingcap/tiflow), is the tool for replicating incremental data of TiKV, which gets change data from TiKV's observer component, with high performance, high available, and scalable.
+
+Besides, **TiKV-CDC** can also provide the ability to connect to other components, e.g. Message Queues, which will help TiKV integrate with customers' data systems, and further extend the applicable scenarios of TiKV.
+
+![rawkv-cdc](../media/rawkv-cdc.png)
+
+### 2. Add Timestamp to Data
+
+As a kind of [Change Data Capture](https://en.wikipedia.org/wiki/Change_data_capture), timestamp (or version) is necessary to indicate which data is changed, and when. While RawKV doesn't have such a thing before, we need to add timestamp to the RawKV data.
+
+#### 2.1 Requirement
+
+Among requests of a key, *Order of Timestamp* must be consistent with sequence of data flush to disk in TiKV.
+
+In general, if request `a` ["happened before"](https://en.wikipedia.org/wiki/Happened-before) `b`, then `Timestamp(a) < Timestamp(b)`. As to cross cluster replication, we provide [Causal Consistency](https://en.wikipedia.org/wiki/Causal_consistency) by keeping order of the timestamp the same as sequence of data flush to disk in TiKV. Downstream systems apply data according to the timestamp order, will always lead to the consistent result.
+
+At the same time, as RawKV doesn't provide cross-rows transaction and snapshot isolation, we treat requests of different keys as concurrent, to reduce complexity and improve performance.
+
+#### 2.2 Timestamp Generation
+
+Timestamp is generated by TiKV internally, to get a better overall performance and client compatibility.
+
+We use [HLC](https://cse.buffalo.edu/tech-reports/2014-04.pdf) method to generate timestamp, which not only is easy to use (by being closed to real world time), but also can capture causality.
+
+##### 2.2.1 Physical Part of Timestamp
+
+Physical part of HLC is acquired from TSO of [PD](https://github.com/tikv/pd).
+
+TSO is a global monotonically increasing timestamp, which help the generation of timestamp be independent to local clock of machine, and be immune to issues such as reverse between machine reboot.
+
+Physical part is refreshed by repeatedly acquiring TSO in a period of `500ms`, to keep it being closed the real world time. And it can tolerate fault of TSO no longer than `30s`, to keep time-related metrics such as RPO reasonable.
```

Comment:
Why `30s`? And will raw write fails if TSO is offline over 30s?

Reply:
En... I think this is not a serious decision. Seems that offline of TSO doesn't break the correctness, except some time-related metrics.
Maybe we can drop this restriction, and report failure metrics. And add some tests on this scenario.

Meanwhile, TiKV-CDC is highly dependent to PD. When PD is offline, the replication is not working.

Reply:
> TiKV-CDC is highly dependent to PD. When PD is offline, the replication is not working. 

Yes, but the write on the main TiKV may still work.

> Maybe we can drop this restriction, and report failure metrics.

Agree

Reply:
Done, PTAL~